### PR TITLE
feat: ARM64 support

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: ISD097

### Overview

Adds support for ARM64 deployment for bare metal arm64 machines (for non-bare metal machines, arm v8.3a or higher is required which supports nested virtualizations. See https://developer.arm.com/documentation/102142/0100/Nested-virtualization)

Link to successful CI runs: https://github.com/yanksyoon/selfhostedtest/actions/runs/7057746518/job/19211975867

### Rationale

To support ARM64 workloads.

### Juju Events Changes

None.

### Module Changes

`charm_state.py` adds a method to determine current architecture.
`runner.py` introduces arm64 yq binary url.

### Library Changes

None.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

Documentation on Charmhub should automatically be updated with the `upload-charm-docs` workflow.

### CODE REVIEW

File: src/runner.py [hunk 1]
```diff
@@ -52,6 +53,9 @@
 
 Snap = NamedTuple("Snap", [("name", str), ("channel", str)])
 
+YQ_BIN_URL_AMD64 = "https://github.com/mikefarah/yq/releases/download/v4.34.1/yq_linux_amd64"
+YQ_BIN_URL_ARM64 = "https://github.com/mikefarah/yq/releases/download/v4.34.1/yq_linux_arm64"
```
Comment (yanksyoon) [inline src/runner.py:57]: Please be reminded that the binaries download is being reworked in another feature branch and will be removed.

File: src/runner.py [hunk 2]
```diff
@@ -498,7 +515,7 @@ def _install_binaries(self, runner_binary: Path) -> None:
         self._wget_install(
             [
                 WgetExecutable(
-                    url="https://github.com/mikefarah/yq/releases/download/v4.34.1/yq_linux_amd64",
+                    url=YQ_BIN_URL_AMD64 if arch is ARCH.X64 else YQ_BIN_URL_ARM64,
                     cmd="yq",
```
Comment (yanksyoon) [inline src/runner.py:519]: The binary download will be refactored in another PR.

File: src/runner_manager.py [hunk 3]
```diff
@@ -193,14 +190,13 @@ def get_latest_runner_bin_url(
         logger.debug("Response of runner binary list: %s", runner_bins)
 
         try:
+            arch = self.config.charm_state.arch.value
```
Comment (yanksyoon) [inline src/runner_manager.py:193]: separate variable assignment to remove "Line break occurred before a binary operator (W503) (https://www.flake8rules.com/rules/W503.html)" error and "Line too long" error.

**Review by yanksyoon — COMMENTED** [review]
Some comments to help review.

**Comment by jdkandersson** [comment]
Could we change the title to not include bare metal since that is more a question of what is supported by the hardware rather than what the charm supports please?

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -31,7 +32,7 @@ def raise_url_error(*args, **kargs):
     raise urllib.error.URLError("mock error")
 
 
-def mock_get_latest_runner_bin_url():
+def mock_get_latest_runner_bin_url(os_name: str = "linux", arch: ARCH = ARCH.X64):
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:35]: Would it be possible and valuable to also have an arm test here?

File: docs/how-to/deploy-on-arm64.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# How to deploy on ARM64
+
+The charm supports deployment on ARM64 machines. However, it should be noted that the ARM64
+deployment currently only supports ARM64 bare-metal machines due to the limitations of
+[nested virtualization on ARM64](https://developer.arm.com/documentation/102142/0100/Nested-virtualization)
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply docs/how-to/deploy-on-arm64.md:5]: Yup, I've asked on the LXD channel to clarify, I can update the docs separately after i'm updated :D

File: docs/how-to/deploy-on-arm64.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+with LXD.
+
+The following uses AWS's [a1.metal](https://aws.amazon.com/blogs/aws/now-available-bare-metal-arm-based-ec2-instances/)
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply docs/how-to/deploy-on-arm64.md]: Fixed, thank you!

File: tests/unit/test_charm.py (hunk 4 above)
Comment (yanksyoon) [reply tests/unit/test_charm.py:35]: Added a test for downloading runner bin!

File: src/charm_state.py [hunk 7]
```diff
@@ -16,8 +16,13 @@
...
 
-ARCHITECTURES_ARM64 = ("aarch64", "arm64")
-ARCHITECTURES_X86 = ("x86_64",)
+ARCHITECTURES_ARM64 = set(
+    (
+        "aarch64",
+        "arm64",
+    )
+)
+ARCHITECTURES_X86 = set(("x86_64",))
```
Comment (jdkandersson) [inline src/charm_state.py:25]: You can declare sets using `{"aarch64", "arm64"}`

**Review by jdkandersson — DISMISSED** [review]

File: tests/unit/test_runner_manager.py [hunk 8]
```diff
@@ -96,27 +98,50 @@ def runner_metrics_fixture(monkeypatch: MonkeyPatch) -> MagicMock:
...
-    assert runner_bin["filename"] == "test_runner_binary"
+    assert runner_bin["architecture"] == arch.value
+    assert runner_bin["download_url"] == download_url
+    assert runner_bin["filename"] == filename
 
 
 def test_get_latest_runner_bin_url_missing_binary(runner_manager: RunnerManager):
     """
-    arrange: Nothing.
+    arrange: Given a mocked GH API client that returns runner binaries.
```
Comment (cbartz) [inline tests/unit/test_runner_manager.py:135]:
```suggestion
    arrange: Given a mocked GH API client that does not return any runner binaries.
```

**Review by cbartz — DISMISSED** [review]
LGTM,  Just one small comment.

**Review by jdkandersson — DISMISSED** [review]

**Review by cbartz — DISMISSED** [review]

File: docs/explanation/arm64.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+virtualizations. 
+
+Furthermore LXD uses QEMU with KVM acceleration options. When run on a nested virtual machine, 
+the following error will occur:
+```
+Error: Failed instance creation: Failed creating instance record: Instance type "virtual-machine"
+is not supported on this server: KVM support is missing (no /dev/kvm)
+```
+
+Therefore, it is necessary that the charm is deployed on a bare metal instance.
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply docs/explanation/arm64.md]: Updated the explanations for why nested virtualizations cannot be done for now.

**Review by jdkandersson — DISMISSED** [review]

**Review by cbartz — DISMISSED** [review]

**Review by cbartz — APPROVED** [review]

**Review by jdkandersson — APPROVED** [review]
